### ISSUE TITLE
Adding inequality/segregation indices

### ISSUE THREAD

**Issue by cdagnino** [QuantEcon/QuantEcon.py#395]
Hi,

Do you think adding ways to calculate inequality/segregation would fit the scope of the quantecon project?
I already have one [in this repository](https://github.com/cdagnino/ReardonSegregationIndex) that I could contribute.

**Comment by jstac** [comment]
@cdagnino My instinct is yes, and the code looks well written.  thanks for offering.  My hesitation is that this code is code is quite isolated from what we have so far.  If it was part of a broader set of code on inequality and segregation measures that would certainly be better.

@mmcky ?

**Comment by mmcky** [comment]
@cdagnino I am keen to diversify the library to include a distributed set of tools that are useful to economists. I will put some time this week into thinking how best we could organise this material, as I have been interested in assembling a ``datalab`` submodule that packages data related routines together.  

My main question is how best to organise the code to sensibly accomodate a variety of different functions that are available from research that may compute the same (or similar) type of output through different methodologies. It would also be good to lay the foundation to allow for a collection of tools for computing inequality and segregation measures (as @jstac suggests). @cdagnino do you use other metrics in your work that would be useful to have in a library?

**Comment by cdagnino** [comment]
Hi,

@jstac @mmcky  I agree 100% with the isolation problem of the code I linked.
If y'all have some thoughts on the interface/syntax for this `datalab` submodule (which I'm guessing would share some ideas from the more general `Quantecon` module) then I can adapt my code to give a first taste of inequality/segregation measures.

I think there are a bunch of inequality and segregation indices that could be interesting to people. Besides the one I coded, a Gini index, a Generalized entropy measure (with Between + Within decompositions) and Duncan Index might be interesting and could be easily made compatible with pandas dataframes. [See this documents for some reference on inequality indices](http://siteresources.worldbank.org/PGLP/Resources/PMch6.pdf)

The "point" estimates should be fine and relatively easy to code. Adding standard errors could be interesting, but more difficult and maybe outside the scope of this package.

**Comment by jstac** [comment]
I think these indices are a great example of what we could usefully code up for economists.  And if there's a number of them we can consider making a subpackage.

@cdagnino Might you be willing to take the lead in developing this subpackage, if we decide to go ahead?  We can probably devote RA resources to helping out, in that case...

**Comment by cdagnino** [comment]
@jstac I'd be interested, but I promised myself not to undertake anything semi-big until I finish my dissertation. I'd be happy to contribute with code and ideas though.

**Comment by jstac** [comment]
@cdagnino Might  you be willing to contribute just a gini index calculator and a Lorenz curve calculator in the short term?  Hopefully it wouldn't take long.

**Comment by cdagnino** [comment]
@jstac Yep! That sounds like doable, at least a simple version without standard errors.

Any suggestions on the implementation, particularly regarding how it should interact to the rest of the QuantEcon codebase?

I'm guessing I could write a `Lorenz` and `Gini` class that takes a `numpy vector` or `pandas Series`. Then these classes can be used to provide numbers for the indices, decompositions and basic plotting.

**Comment by cdagnino** [comment]
@jstac Took a first try [here](https://github.com/cdagnino/inequality_measures_gym/blob/master/Lorenz%20and%20Gini.ipynb)

Haven't implemented much, but I figured out the issue you had with the Pareto curve sampling and the Gini coefficient.
Turns out `numpy` doesn't have a real Pareto distribution, but rather a Pareto II or Lomax!
Once you "rescale", your graph works fine.

Stay tuned.

**Comment by jstac** [comment]
@cdagnino  Nice, thanks!  I guess you could use the same idea for the `gini` test --- at those extreme distributions the coefficient should be close to zero or close to one.

And thanks for the fix vis-a-vis the Pareto distribution!

If you get a chance to finish the tests then please submit this code as a PR and @mmcky (the package maintainer) will help get it merged.

We're just about to launch this site:

http://notes.quantecon.org/

and your notebook could be posted there to show people how to use the functions, if you're willing.

**Comment by jstac** [comment]
@rebekahanne Might you have time to help get this merged?

**Comment by jstac** [comment]
@rebekahanne Great, thanks.  I'd appreciate it.  The code is there already and we want to implement this functionality just as functions with corresponding tests.  This is the main QuantEcon Python code library and @mmcky is the current gatekeeper.  You'll need to fork the repository, collect together the code and tests in a branch and them submit a pull request.  If you browse the repository you'll see how functions are combined with tests. @natashawatkins or @mmcky or @QBatista could help if you have questions.

**Comment by rebekahanne** [comment]
@jstac All right -- will do!

**Comment by cdagnino** [comment]
@jstac @rebekahanne Apologies for my silence. I've made the changes on #414 . Hopefully I can make a pull with that.

**Comment by rebekahanne** [comment]
@cdagnino I'll write the tests etc. this weekend so that we can merge the code soon.

**Comment by jstac** [comment]
Thanks guys.  @rebekahanne It looks like there are some tests in the branch that @cdagnino references above.  But it would be great if you could review the code and help it move forward until merged.  It would also be nice to have a lecture that shows off this code once it is merged.  @rebekahanne, perhaps we could discuss this again later.

**Comment by cdagnino** [comment]
@rebekahanne I pushed some new code and tests yesterday on #414 . If you wanna review it then hopefully I can do a pull request on that

**Comment by cdagnino** [comment]
I can code the Duncan Index and the Reardon Index next. 

For the lecture notes, I think a nice application would be to use (comparable) income data for a few countries, but a quick google search didn't show up much.

**Comment by rebekahanne** [comment]
@cdagnino Thank you! I can talk to @jstac about how we would want to develop a lecture.

**Comment by jstac** [comment]
Thanks guys, sounds great.  

As a start we can illustrate the indices with a notebook here with some basic examples: http://notes.quantecon.org/.  Perhaps you two might be willing to collaborate and put your names on that?  Even examples with simulated data will be useful to readers.

I have already written code for a lecture with the Lorenz curve function and the Gini coefficient function.  I'll put the whole thing together when I have a bit of time.

**Comment by rebekahanne** [comment]
That sounds like a good plan.

@cdagnino , if you're willing, we can coordinate to put together a notebook with examples/explanations of the code etc.

**Comment by cdagnino** [comment]
@rebekahanne Sure! Do y'all use gitter / slack or something else for a smoother interaction?

**Comment by natashawatkins** [comment]
I can add you to the slack. What’s your email?

**Comment by cdagnino** [comment]
cdagnino at gmail

**Comment by natashawatkins** [comment]
Hi @rebekahanne and @cdagnino - is the notebook almost ready to go? I think a bit more explanation could be added, and then we should publish it to notes.quantecon.org and make note of the new functions in QuantEcon.py. Is there anythin else you are planning to add?

**Comment by natashawatkins** [comment]
It would also be great to add more inequality measures to QuantEcon.py in the near future

**Comment by rebekahanne** [comment]
I got fairly distracted exams, so we still need to add a bit more. The data you gave us actually has many years available, so it might interesting to look at a time series of Gino coefficients and Lorenz curves etc.

**Comment by jstac** [comment]
@rebekahanne Sounds great.  It would be fascinating to see how they both evolve.

The Lorenz curves could be bunched together into a 3D surface, with time on one axis, to show how they have changed, 

@cdagnino You mentioned contributing [this code](https://github.com/cdagnino/ReardonSegregationIndex) a while back.  Are you still willing?  What else should we add?  How about the Shorrocks segregation index?

**Comment by rebekahanne** [comment]
@jstac That's a great idea, thank you! My first thought was just to plot the contours, but I'll try the 3D surface, too, to see if this helps clearly display change over time.

**Comment by natashawatkins** [comment]
Hey @rebekahanne do you maybe have time to finish this off? I just tried plotting over multiple years but it doesn't appear that there's much variation

We should also add an example of using the Gini coefficient and the Shorrucks index. Maybe we can find some other countries data... last time I found it quite hard but maybe you are more familiar with the US SCF?

<img width="590" alt="screen shot 2019-03-04 at 7 19 40 pm" src="https://user-images.githubusercontent.com/21038129/53719560-91a32600-3eb2-11e9-8053-bd827df52311.png">
<img width="518" alt="screen shot 2019-03-04 at 7 19 44 pm" src="https://user-images.githubusercontent.com/21038129/53719561-91a32600-3eb2-11e9-9730-dda5932340ba.png">

**Comment by sidd1996** [comment]
Hi all can you please let me know exactly what is the open issue here now ? Thank you

**Comment by jstac** [comment]
I think this issue can probably be closed.  (@natashawatkins , glancing it the thread the above, is it the case that you already have a notebook illustrating QE's inequality tools https://quanteconpy.readthedocs.io/en/latest/tools/inequality.html?  If so, would you mind to post it to notes.quantecon.org?  If not then please disregard.)

@sidd1996 , thanks for your interest. if you would like to contribute, please email me, let me know your background, and I'll see if we have something suitable. https://johnstachurski.net/
